### ISSUE TITLE
more aggressive configuration abstraction wanted

### ISSUE THREAD

**Issue by grosu** [kframework/k-legacy#2075]
We've discussed this in the past and probably there is some old issue about it, but I'm adding this new issue here as I revise the tutorial, to make sure that we keep track of it.  Hopefully some day soon we will be able to revise the configuration abstraction algorithm and then this will be useful.

So, instead of writing this:

```
rule <k> spawn S => . ...</k> <env> Rho </env>
     (. => <thread>... <k> S </k> <env> Rho </env> ...</thread>)
```

we'd like to write this:

```
rule <k> spawn S => . ...</k> <env> Rho </env>
     (. => <k> S </k> <env> Rho </env>)
```

See https://github.com/kframework/k/tree/master/k-distribution/tutorial/1_k/4_imp%2B%2B/lesson_6 for details, and update it once this works.

**Comment by cos** [comment]
For the record, I find the second version more cryptic. Unless someone knows the context, it is not obvious a new thread is spawn. E.g., I'm not sure I would have understood it immediately if I didn't have the first version.

This note is not about the power of our configuration concretization mechanism. We could have something powerful enough to support the second but choose to write it as the first.

**Comment by grosu** [comment]
Exactly.  It is one's choice how to write these things.  Our duty is to provide a powerful structure inference mechanism, based on solid principles and backed by a serious theorem (same like type inference in higher-order languages).  Then users decide themselves how they want to use it.  For example, many users of functional languages tag variables with their types "just in case".

Modularity extremists may prefer the second variant above because their rule does not depend on the particular name of the `<thread/>` cell.  Tomorrow somebody can change it into `<process/>` or `<agent/>` and everything still works.
